src/gtk/spinctrl.cpp

```diff
@@ -364,26 +365,32 @@ wxSize wxSpinCtrlGTKBase::DoGetSizeFromTextSize(int xlen, int ylen) const
 {
     wxASSERT_MSG( m_widget, wxS("GetSizeFromTextSize called before creation") );
 
+    // This is a bit stupid as we typically compute xlen by measuring some
+    // string of digits in the first place, but there doesn't seem to be
+    // anything better to do (unless we add some GetSizeFromNumberOfDigits()).
+    const double widthDigit = GetTextExtent("0123456789").GetWidth() / 10.0;
+    const int numDigits = wxRound(xlen / widthDigit);
+
     const gint widthChars = gtk_entry_get_width_chars(GTK_ENTRY(m_widget));
-    gtk_entry_set_width_chars(GTK_ENTRY(m_widget), 0);
-#if GTK_CHECK_VERSION(3,12,0)
-    gint maxWidthChars = 0;
-    if ( gtk_check_version(3,12,0) == NULL )
-    {
-        maxWidthChars = gtk_entry_get_max_width_chars(GTK_ENTRY(m_widget));
-        gtk_entry_set_max_width_chars(GTK_ENTRY(m_widget), 0);
-    }
-#endif // GTK+ 3.12+
+    gtk_entry_set_width_chars(GTK_ENTRY(m_widget), numDigits);
 
-    wxSize totalS = GTKGetPreferredSize(m_widget);
+    wxSize tsize = GTKGetPreferredSize(m_widget);
 
-#if GTK_CHECK_VERSION(3,12,0)
-    if ( gtk_check_version(3,12,0) == NULL )
-        gtk_entry_set_max_width_chars(GTK_ENTRY(m_widget), maxWidthChars);
-#endif // GTK+ 3.12+
     gtk_entry_set_width_chars(GTK_ENTRY(m_widget), widthChars);
 
-    wxSize tsize(xlen + totalS.x, totalS.y);
+    // Somehow, gtk_widget_get_preferred_size() doesn't account for the borders
```

Comment:
If this were true, there would be a problem with the height as well. What is happening here is `widthChars == -1`, so the actual preferred size is being computed using the upper and lower bounds of the range (see the GTK3 private function `gtk_spin_button_get_text_width()`). It seems that this is sometimes turning out to be just a couple pixels bigger than the preferred size we're getting for `numDigits`.

Reply:
Why is `widthChars == -1` though when we explicitly set it to 0 in the current master and `numDigits` in this PR?

Also, it does happen with the height too (preferred height is 36, allocated height is 34), but I just ignored it because the height is much bigger than the font height anyhow, so it doesn't matter. I probably shouldn't be lazy and do it for the height as well.

This being said, I'm still not sure about what's really going on here and if this has anything to do with `wxSpinButton::DoGetBestSize()`, where I don't understand why is it removing padding from the best width neither...

Reply:
> Why is widthChars == -1 though when we explicitly set it to 0 in the current master and numDigits in this PR?

I meant literally the `widthChars` variable. Try this:
```diff
--- src/gtk/spinctrl.cpp-	2019-11-12 07:49:10.089190178 -0800
+++ src/gtk/spinctrl.cpp	2019-11-12 07:49:39.890547001 -0800
@@ -375,8 +375,10 @@
     gtk_entry_set_width_chars(GTK_ENTRY(m_widget), numDigits);
 
     wxSize tsize = GTKGetPreferredSize(m_widget);
+fprintf(stderr, "%d", tsize.x);
 
     gtk_entry_set_width_chars(GTK_ENTRY(m_widget), widthChars);
+fprintf(stderr, " %d\n", GTKGetPreferredSize(m_widget).x);
 
     // Somehow, gtk_widget_get_preferred_size() doesn't account for the borders
     // with this widget, and if we use it as is, we get GTK errors when
```
> Also, it does happen with the height too (preferred height is 36, allocated height is 34)

I don't see any evidence of this. The reported preferred height is 34, and the control is perfectly happy being sized to that, but if you size it to 33, it complains.
>wxSpinButton::DoGetBestSize(), where I don't understand why is it removing padding from the best width neither...

Because the padding is for the text part of the control, which we don't want to show for `wxSpinButton`, see also 135e3875e2.

Reply:
OK, I see, thanks. I've actually tried calling `GTKGetPreferredSize()` with different char width values under debugger, so I did see this, but didn't think it through.

However I am still not sure what should be done about this: should we just call `gtk_entry_set_width_chars()` in ctor/`DoSetRange()`? I.e. instead of leaving it as default -1 always set it to the actually needed value?

BTW, am I correct in thinking that there is no difference between `PROP_WIDTH_CHARS` and `PROP_MAX_WIDTH_CHARS` in our case? AFAICS the former is used to compute the min size and the latter -- the "natural" size, but they're the same for `wxSpinCtrl`.

Reply:
> should we just call gtk_entry_set_width_chars() in ctor/DoSetRange()?

Probably.
> am I correct in thinking that there is no difference between PROP_WIDTH_CHARS and PROP_MAX_WIDTH_CHARS in our case?

Also probably. As long as nothing sets `max-width-chars`.